### PR TITLE
feature(product): creation of entity Ingredient

### PR DESCRIPTION
**Pull request by rhonniel**
Ingredient is the material of product, for example orange,milk or weed. For now is just a name and the record can be inactivated.

### CODE REVIEW

File: src/com/mercado/model/product/Ingredient.java [hunk 1]
```diff
@@ -0,0 +1,31 @@
+package com.mercado.model.product;
+
+public class Ingredient {
+    private int id;
```
Comment (eapg) [inline src/com/mercado/model/product/Ingredient.java]: Avoid using primitive data structures.

File: src/com/mercado/model/product/Ingredient.java [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+    private String name;
+    private boolean isActive;
```
Comment (eapg) [inline src/com/mercado/model/product/Ingredient.java]: create a `BaseEntity` to be inherit for all your entities with `status`, `createdBy`, `createdate`, `updatedBy`, `updatedDate` and all the common things in your entities.

File: src/com/mercado/model/product/Ingredient.java (hunk 1 above)
Comment (eapg) [inline src/com/mercado/model/product/Ingredient.java]: the `id` could be in your `baseEntity` as every entity has to have an `id`

File: src/com/mercado/model/product/Ingredient.java [hunk 3]
```diff
@@ -0,0 +1,31 @@
+package com.mercado.model.product;
+
```
Comment (eapg) [inline src/com/mercado/model/product/Ingredient.java:2]: add constructor for your entity.

**Review by eapg — CHANGES_REQUESTED** [review]
Also add `equals()` and `hasCode()`

File: src/com/mercado/model/product/Ingredient.java (hunk 1 above)
Comment (rhonniel) [reply src/com/mercado/model/product/Ingredient.java]: Ok , but why?, explain your review

File: src/com/mercado/model/product/Ingredient.java (hunk 3 above)
Comment (rhonniel) [reply src/com/mercado/model/product/Ingredient.java:2]: ok two thinks

-  there is a constructor ,  the default
- i dont need another for now, or explain me why a need another  in this case, in this PR not in the future

File: src/com/mercado/model/product/Ingredient.java (hunk 2 above)
Comment (rhonniel) [reply src/com/mercado/model/product/Ingredient.java]: - Ok i like the base entity  sound  good
- But i dont really add status , status sucks, why? i  just need two states (available , unavailable)
- createdBy, createdate, updatedBy, updatedDate....  for now I dont wanna manege historical data, and existe better forms, for example save  all updates in a new entity.  Also createdBy is relative to User entity , this PR is about Product

File: src/com/mercado/model/product/Ingredient.java (hunk 1 above)
Comment (rhonniel) [reply src/com/mercado/model/product/Ingredient.java]: - yes , thanks

File: src/com/mercado/model/product/Ingredient.java (hunk 2 above)
Comment (reynoldmorel) [reply src/com/mercado/model/product/Ingredient.java]: I think having a status that could be extended to multiple statuses is a good idea. You may deal with incative records in the future. Why not having just one field to handle that?

**Review by reynoldmorel — CHANGES_REQUESTED** [review]

File: src/com/mercado/model/product/Ingredient.java (hunk 2 above)
Comment (rhonniel) [reply src/com/mercado/model/product/Ingredient.java]: yes inactive is one and active is the other, only two. I can't figuret out another state. But ok is exists 3 i need to add Status entity, a espesific one  for baseEntities.

File: src/com/mercado/model/product/Ingredient.java (hunk 1 above)
Comment (reynoldmorel) [reply src/com/mercado/model/product/Ingredient.java]: Java included a lot of features on its core that you are not taking advantage of, just because of primitive values. 

1. You are preventing your code from using latest java features in the first try of your code.
2. Java implementation knows that there could be performance issues with their implementation, hence they have imnprove the compiler to box and unbox primitive values as well was parsing them.
3. Taking advantage of other java features example Collections, Streaming, Futures, Optionals and Functions might make you create extra verbose code.

In summary, use primitive only if the implementation of the java core is affecting you performance

File: src/com/mercado/model/product/Ingredient.java (hunk 3 above)
Comment (reynoldmorel) [reply src/com/mercado/model/product/Ingredient.java:2]: It is good to be as much explicit as you can so other developers  from other projects can easily see your code.

If you are not going to add any constructor then at least add the default one. Not adding the constructor can lead to confusion when other devs try to add custom constructors to your entities, so you make life easier for them

File: src/com/mercado/model/product/Ingredient.java (hunk 1 above)
Comment (rhonniel) [reply src/com/mercado/model/product/Ingredient.java]: ok, thanks

File: src/com/mercado/model/product/Ingredient.java (hunk 2 above)
Comment (rhonniel) [reply src/com/mercado/model/product/Ingredient.java]: ok , i just surrender in this point

File: src/com/mercado/model/product/Ingredient.java (hunk 3 above)
Comment (rhonniel) [reply src/com/mercado/model/product/Ingredient.java:2]: "add custom constructors to your entities, so you make life easier for them" Exactly, but now , what is the point, the case, the proyect dont have any logic, any action. I dont want create a costructor for nothing, just becouse look good.

Comment (rhonniel) [reply src/com/mercado/model/product/Ingredient.java:2]: ok just ignore me i'll add constructors

File: src/com/mercado/model/BaseEntity.java [hunk 4]
```diff
@@ -0,0 +1,44 @@
+package com.mercado.model;
+
+import java.util.Objects;
+
+public abstract class BaseEntity {
+    private Integer id;
```
Comment (reynoldmorel) [inline src/com/mercado/model/BaseEntity.java:8]: What do you think about using 
`Long` for ids?

Comment (rhonniel) [reply src/com/mercado/model/BaseEntity.java:8]: I think 2,147,483,647 records is enough for mayority  cases, but only in the scenario where we use something like a chat or record for every fly born in the world would we need
 more space.

Comment (reynoldmorel) [reply src/com/mercado/model/BaseEntity.java:8]: Fair enough

File: src/com/mercado/model/BaseEntity.java [hunk 5]
```diff
@@ -0,0 +1,44 @@
...
+
+    public Integer getId() {
+        return id;
+    }
+
+    public void setId(Integer id) {
+        this.id = id;
+    }
+
+    public String getName() {
```
Comment (reynoldmorel) [inline src/com/mercado/model/BaseEntity.java]: Could you move name to the ingredient class? Not all entities are supposed to have a name.

**Review by reynoldmorel — CHANGES_REQUESTED** [review]

**Review by reynoldmorel — APPROVED** [review]

**Review by eapg — APPROVED** [review]
